### PR TITLE
[WIP] mongoose implementation

### PR DESCRIPTION
**Pull request by yharaskrik**
Basic mongoose implementation, working branch.

### CODE REVIEW

**Comment by Bnaya** [comment]
Awesome!

**Comment by yharaskrik** [comment]
Basic joins are now possible via the crud-mongoose package. 

Requirements for joins to work:

Mongoose schema must be setup to use virtuals
Virtual for the join that is requested must be setup, see here: https://mongoosejs.com/docs/populate.html

**Comment by Bnaya** [comment]
if this feature is mostly ready, would be awesome to have it published as separate package until merge, as canary or so,
For users to provide feedback, etc

Thanks!

**Comment by yharaskrik** [comment]
Hey @bnaya,

There is still some unit testing I would like to write but I could publish it as a separate package for now! I will update this soon! Just heading home for Christmas but would like to get that done soon!

**Comment by yharaskrik** [comment]
I just published a 0.0.1 version to npm here

`https://www.npmjs.com/package/nest-crud-mongoose`

I need to update the readme to reflect mongoose instead of typeORM which I will try and get done today.

Feel free to comment on this issue with any questions or concerns. I will comment again when the docs after been updated.

**Comment by yharaskrik** [comment]
Updated.

**Comment by yharaskrik** [comment]
@Bnaya have you used the package at all?

**Comment by Bnaya** [comment]
I've created this project template based on your integration tests code + in memory mongo
https://codesandbox.io/s/github/Bnaya/demo-mongoose-implementation-nest-crud/

File: integration/crud-mongoose/comments/comments.module.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { Module } from '@nestjs/common';
+
+@Module({
+  controllers: [
+    Comment
+  ]
+})
```
Comment (Bnaya) [inline integration/crud-mongoose/comments/comments.module.ts]: missing class

File: integration/crud-mongoose/posts/posts.controller.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import { Crud, CrudController } from '@nestjsx/crud';
+import { Post } from './post.entity';
+import { PostsService } from './posts.service';
+
+@Crud({
+  model: {
+    type: Post,
+  },
+})
+@Controller('posts')
```
Comment (Bnaya) [inline integration/crud-mongoose/posts/posts.controller.ts]: got error:
```
Class 'PostsController' incorrectly implements interface 'CrudController<Post>'.
  Types of property 'service' are incompatible.
    Type 'PostsService' is not assignable to type 'CrudService<Post>'.
      Property 'getTake' is private in type 'PostsService' but not in type 'CrudService<Post>'
```
Same the rest of controllers 

// @ts-ignore made it pass and work

Comment (yharaskrik) [reply integration/crud-mongoose/posts/posts.controller.ts]: Thanks, will look at that

File: integration/crud-mongoose/comments/comments.module.ts (hunk 1 above)
Comment (yharaskrik) [reply integration/crud-mongoose/comments/comments.module.ts]: What do you mean?

Comment (Bnaya) [reply integration/crud-mongoose/comments/comments.module.ts]: this part of the file is missing:
```
export class CommentsModule {
}

```

File: integration/crud-mongoose/posts/posts.controller.ts (hunk 2 above)
Comment (yharaskrik) [reply integration/crud-mongoose/posts/posts.controller.ts]: How do you cause this to happen? When I build or run the integration tests that does not show up for me.

Also `CrudService<T>` does not have a function `getTake<T>` from what I can tell.

Comment (Bnaya) [reply integration/crud-mongoose/posts/posts.controller.ts]: Maybe its next/crud versions issue
look at:
https://codesandbox.io/s/bitter-dust-cgo0r?fontsize=14&hidenavigation=1&module=%2Fsrc%2Fusers%2Fusers.controller.ts&theme=dark&view=editor

Comment (yharaskrik) [reply integration/crud-mongoose/posts/posts.controller.ts]: Yes that is what it looks like. I will update the crud package and make sure it still works.

**Comment by yharaskrik** [comment]
I have rebased my code onto what is released on nestjsx/crud but am having issues with the mongoose documents not being converted to POJOs now, the request just returns a JSON object of the document not the value of the document itself. Not sure what changed with this but toJSON isn't being called correctly. I will try and figure it out soon.

**Comment by yharaskrik** [comment]
@zMotivat0r is there something that would have changed in 4.3.x that would have caused the requests to not toJSON the mongoose Documents? All of my tests passed before but now the Documents are returned as they are represented in memory instead of the value of the document itself.

I had to remove all references of authFilter and the getTake and getSkip methods plus I had to remove the code that prefixed the operators with a $ as they before the operators were just `eq` or `lt`, now they are `$eq`. None of those changes should have changed the behavior around the return objects. I can try and force them to all return as `lean()` so they are plain JSON objects but then the `_id` field is returned as an object buffer as well whereas before it was being returned as the string representation of the buffer (as it should) so that the front end gets back the mongo ObjectID for the record.

Thanks for any input you have..

**Comment by zMotivat0r** [comment]
`getTake` and `getSkip` have been removed to the abstract class. 
Regarding the response - I've added the `CrudResponseInterceptor`, please see in `crud/interceptors` - there is `classToPlain` calls

**Comment by yharaskrik** [comment]
Yup, noticed those methods were added to abstract thus were not needed, that is nice, the intercetpor is probably what is doing it! Thanks I will take a look.

**Comment by yharaskrik** [comment]
So I am going to need to spend some more time on this, for now the interceptors need to be turned off like in my spec files for it to work because of the way mongoose serializes the `Document` object.

**Comment by yharaskrik** [comment]
`0.0.5` released, has support for using the nestjsx/crud interceptors and custom serialization now.

**Comment by lafeuil** [comment]
I have tried your `MongooseCrudService` with the `CrudAuth` decorator. The `MongooseCrudService` does not include the filter in the query. After comparing with the `OrmCrudService`, this uses the `parsed.search` variable previously calculate to construct the query.

I think, in the `createBuilder` function, you should define the `defaultSearch` variable by calling the `queryFilterToSearch` function with `parsed.search` : 

```typescript
  public async createBuilder<K>(
    fn: (...args) => DocumentQuery<K, T>,
    parsed: ParsedRequestParams,
    options: CrudRequestOptions,
    many = true,
  ): Promise<{ builder: DocumentQuery<K, T>; take?: number; skip?: number }> {
    // get select fields
    const select = this.getSelect(parsed, options.query);
    // default search condition
    const defaultSearch = this.queryFilterToSearch(parsed.search);

    const builder = fn(defaultSearch);
    // select fields
    builder.select(select);
```

**Comment by yharaskrik** [comment]
Hey @lafeuil, thanks for letting me know! I will take a look at that today, could you provide me with the code for the Crud controller you were using in the example you provided?

**Comment by lafeuil** [comment]
For example, if you want to filter the posts by the user that is authenticated :

```typescript
@Crud({
  model: {
    type: Post,
  },
})
@CrudAuth({
  filter: (user) => {
    return { userId: user.id };
  },
@Controller('posts')
export class PostsController implements CrudController<Post> {
...
}
```
Example in the integration TypeOrm : 
https://github.com/nestjsx/crud/blob/bf54fe8b1d89e89bf5bdd244f0dbb311b7052fc7/integration/crud-typeorm/projects/my-projects.controller.ts#L28-L31

You can login a user automatically with the AuthGuard in the integration TypeOrm : 
https://github.com/nestjsx/crud/blob/bf54fe8b1d89e89bf5bdd244f0dbb311b7052fc7/integration/crud-typeorm/auth.guard.ts#L1-L16

**Comment by yharaskrik** [comment]
Awesome thank you! Will take a look at that today!

File: packages/crud-mongoose/src/mongoose-crud.service.ts [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+
+    const { builder, take, skip } = await this.createBuilder<T[]>(
+      this.find,
+      parsed,
+      options,
+    );
+
+    if (this.decidePagination(parsed, options)) {
+      const data = await builder;
+      const total = await this.count({});
```
Comment (lafeuil) [inline packages/crud-mongoose/src/mongoose-crud.service.ts]:
```suggestion
      const total = await this.count(parsed.search);
```

File: packages/crud-mongoose/src/mongoose-crud.service.ts [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+  public decidePagination(
+    parsed: ParsedRequestParams,
+    options: CrudRequestOptions,
+  ): boolean {
+    return (
+      (Number.isFinite(parsed.page) || Number.isFinite(parsed.offset)) &&
+      !!this.getTake(parsed, options.query)
+    );
+  }
+
```
Comment (lafeuil) [inline packages/crud-mongoose/src/mongoose-crud.service.ts]: Remove this to use the implementation in the parent class

Comment (yharaskrik) [reply packages/crud-mongoose/src/mongoose-crud.service.ts]: Removed, will push it soon.

File: packages/crud-mongoose/src/mongoose-crud.service.ts (hunk 3 above)
Comment (yharaskrik) [reply packages/crud-mongoose/src/mongoose-crud.service.ts]: Correct, I will switch this but am having issues with `parsed.search` being `{ '$and': [ undefined, {} ] }` which is throwing errors, trying to figure out why it is not being parsed correctly.

**Comment by yharaskrik** [comment]
Hey @zMotivat0r ,

Just wondering if you would know what is going on here as it is the crud-request code causing it.

When I run this test in the mongoose package:

```
it('should return an entities with limit and page', (done) => {
        const query = qb
          .setLimit(3)
          .setPage(1)
          .sortBy({ field: '_id', order: 'DESC' })
          .query();
        return request(server)
          .get('/users')
          .query(query)
          .end((_, res) => {
            expect(res.status).toBe(200);
            expect(res.body.data.length).toBe(3);
            expect(res.body.count).toBe(3);
            expect(res.body.total).toBe(9);
            expect(res.body.page).toBe(1);
            expect(res.body.pageCount).toBe(3);
            done();
          });
      });
```

The `parsed.search` property on the Request object is `{ '$and': [ undefined, {} ] }`. This throws an error with mongoose when trying to use it to query on `model.count(parsed.search)`.

For the time being I wrote some code to prevent invalid values to be used for building the query here:

```
parsed.search = Object.keys(parsed.search).reduce((prev: { [key: string]: any[] }, cur: string) => {
      const queries = parsed.search[cur].filter(query => isObjectFull(query));
      if (isArrayFull(queries)) {
        prev[cur] = queries;
      }

      return prev;
    }, {});
```

But would like to just be able to use `parsed.search` and have the crud-request code not provide `undefined` or `{}` in the `$and` key of the `parsed.search` param.

**Comment by yharaskrik** [comment]
> For example, if you want to filter the posts by the user that is authenticated :
> 
> ```ts
> @Crud({
>   model: {
>     type: Post,
>   },
> })
> @CrudAuth({
>   filter: (user) => {
>     return { userId: user.id };
>   },
> @Controller('posts')
> export class PostsController implements CrudController<Post> {
> ...
> }
> ```
> 
> Example in the integration TypeOrm :
> 
> https://github.com/nestjsx/crud/blob/bf54fe8b1d89e89bf5bdd244f0dbb311b7052fc7/integration/crud-typeorm/projects/my-projects.controller.ts#L28-L31
> 
> You can login a user automatically with the AuthGuard in the integration TypeOrm :
> 
> https://github.com/nestjsx/crud/blob/bf54fe8b1d89e89bf5bdd244f0dbb311b7052fc7/integration/crud-typeorm/auth.guard.ts#L1-L16

Hey @lafeuil , my apologies I have not had time to look at this yet, will try and look tomorrow if I can find the time.

**Comment by 9173860** [comment]
@yharaskrik I found search or filter in query is not working and seems not been implemented.

Is this related to the following issue?

> Hey @zMotivat0r ,
> 
> Just wondering if you would know what is going on here as it is the crud-request code causing it.
> 
> When I run this test in the mongoose package:
> 
> ```
> it('should return an entities with limit and page', (done) => {
>         const query = qb
>           .setLimit(3)
>           .setPage(1)
>           .sortBy({ field: '_id', order: 'DESC' })
>           .query();
>         return request(server)
>           .get('/users')
>           .query(query)
>           .end((_, res) => {
>             expect(res.status).toBe(200);
>             expect(res.body.data.length).toBe(3);
>             expect(res.body.count).toBe(3);
>             expect(res.body.total).toBe(9);
>             expect(res.body.page).toBe(1);
>             expect(res.body.pageCount).toBe(3);
>             done();
>           });
>       });
> ```
> 
> The `parsed.search` property on the Request object is `{ '$and': [ undefined, {} ] }`. This throws an error with mongoose when trying to use it to query on `model.count(parsed.search)`.
> 
> For the time being I wrote some code to prevent invalid values to be used for building the query here:
> 
> ```
> parsed.search = Object.keys(parsed.search).reduce((prev: { [key: string]: any[] }, cur: string) => {
>       const queries = parsed.search[cur].filter(query => isObjectFull(query));
>       if (isArrayFull(queries)) {
>         prev[cur] = queries;
>       }
> 
>       return prev;
>     }, {});
> ```
> 
> But would like to just be able to use `parsed.search` and have the crud-request code not provide `undefined` or `{}` in the `$and` key of the `parsed.search` param.

**Comment by yharaskrik** [comment]
@9173860 I believe in the version that was merged into v5 did have search and filter working, the comment you reference is talking about some code that i haven't pushed yet. 

Could you open a new issue with more info about them problem you are having? Thanks a lot.

**Comment by 9173860** [comment]
> @9173860 I believe in the version that was merged into v5 did have search and filter working, the comment you reference is talking about some code that i haven't pushed yet.
> 
> Could you open a new issue with more info about them problem you are having? Thanks a lot.

Thanks for the reply. I've opened #435 for reporting it with a demo.

**Pull request merged**